### PR TITLE
MGMT-7472: Create Agent CR without Cluster

### PR DESCRIPTION
**Pull request by rollandf**
# Assisted Pull Request

## Description

In case the InfranEnv is not associated to a Cluster, the Agent CR
should be created without a reference to a ClusterDeployment.

Also, make the ClusterDeployment reference in Agent CR optional.

Signed-off-by: Fred Rolland <frolland@redhat.com>

## List all the issues related to this PR

- [x] New Feature
- [ ] Bug fix
- [ ] Tests
- [ ] Documentation
- [ ] CI/CD 

## What environments does this code impact?

- [ ] Automation (CI, tools, etc)
- [ ] Cloud
- [x] Operator Managed Deployments
- [ ] None

## How was this code tested?

- [ ] assisted-test-infra environment
- [ ] dev-scripts environment
- [ ] Reviewer's test appreciated
- [x] Waiting for CI to do a full test run
- [ ] Manual (Elaborate on how it was tested)
- [ ] No tests needed

## Assignees

/cc @filanov 
/cc @yevgeny-shnaidman 
/cc @nmagnezi 

## Checklist

- [x] Title and description added to both, commit and PR.
- [x] Relevant issues have been associated (see [CONTRIBUTING] guide)
- [x] Reviewers have been listed
- [x] This change does not require a documentation update (docstring, `docs`, README, etc)
- [x] Does this change include unit-tests (note that code changes require unit-tests)

## Reviewers Checklist

- [ ] Are the title and description (in both PR and commit) meaningful and clear?
- [ ] Is there a bug required (and linked) for this change?
- [ ] Should this PR be backported?

[Kubernetes community documentation]: https://github.com/kubernetes/community/blob/master/contributors/guide/pull-requests.md#commit-message-guidelines
[CONTRIBUTING]: https://github.com/openshift/assisted-service/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: internal/controller/controllers/crd_utils.go [hunk 1]
```diff
@@ -55,48 +63,51 @@ func (u *CRDUtils) CreateAgentCR(ctx context.Context, log logrus.FieldLogger, ho
...
 		// fetch previous by KubeKey
 		h, err2 := u.hostApi.GetHostByKubeKey(key)
 		if err2 != nil && !errors.Is(err2, gorm.ErrRecordNotFound) {
 			return errors.Wrapf(err2, "Failed to GetHostByKubeKey Name: %s Namespace: %s", key.Name, key.Namespace)
 		}
 
 		if err2 == nil {
-			if *h.ClusterID == *clusterID {
+			if cluster != nil && h.ClusterID != nil && *h.ClusterID == *cluster.ID {
 				log.Infof("Agent CR %s already exists, same cluster %s", hostId, h.ClusterID)
```
Comment (yevgeny-shnaidman) [inline internal/controller/controllers/crd_utils.go:111]: if the host belong to infra-env only (if it is in pool) then does not it means that we will deregister it at line 115?

Comment (rollandf) [reply internal/controller/controllers/crd_utils.go:111]: Good catch. Added a test and fix.

**Pull request merged**